File: importado-do-antigo/ALGORITMOS-E-ESTRUTURAS-DE-DADOS-II (PT2)/#Teorico/Códigos de Estudos/3 - Estruturas de Dados Lineares e Flexíveis/src/ListaDuplamenteEncadeada.java
public class ListaDuplamenteEncadeada {
    private Celula primeiro;
    private Celula ultimo;
    private int tamanho;

    public ListaDuplamenteEncadeada() {
        Celula sentinela = new Celula();

        this.primeiro = sentinela;
        this.ultimo = sentinela;
        this.tamanho = 0;
    }

    public boolean listaVazia() {
        return (this.primeiro == this.ultimo);
    }

    public void inserirFinal(Inteiro novo) {
        Celula novaCelula = new Celula(novo);

        this.ultimo.setProximo(novaCelula);
        novaCelula.setAnterior(this.ultimo);

        this.ultimo = novaCelula;
        this.tamanho++;
    }

    public Inteiro removerFinal() throws Exception {
        Celula removida, penultima;

        if (!listaVazia()) {
            removida = this.ultimo;

            penultima = this.ultimo.getAnterior();
            penultima.setProximo(null);
            removida.setAnterior(null);

            this.ultimo = penultima;
            this.tamanho--;

            return (removida.getItem());
        } else
            throw new Exception("Não foi possível remover o último item da lista: a lista está vazia!");
    }

    public void imprimir() throws Exception {
        Celula aux;

        if (!listaVazia()) {
            System.out.println("Conteúdo da lista:");
            aux = this.primeiro.getProximo();
            while (aux != null) {
                aux.getItem().imprimir();
                aux = aux.getProximo();
            }
        } else
            throw new Exception("Não foi possível imprimir o conteúdo da lista: a lista está vazia!");
    }
}

File: importado-do-antigo/ALGORITMOS-E-ESTRUTURAS-DE-DADOS-II (PT3)/Algoritmos e Estruturas de Dados II - TEO/TAD - Tipo Abstrato de Dados/ListaDuplamenteEncadeada/ListaDuplamenteEncadeada/ListaDuplamenteEncadeada.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ListaDuplamenteEncadeada{
	class ListaDuplamenteEncadeada{
		private Celula primeiro;
		private Celula ultimo;
		private int tamanho;

		public ListaDuplamenteEncadeada(){
			Celula sentinela;

			sentinela = new Celula();

			primeiro = sentinela;
			ultimo = sentinela;

			tamanho = 0;
		}

		public Boolean listaVazia(){

			Boolean resp;

			if (primeiro == ultimo)
				resp = true;
			else
				resp = false;

			return resp;
		}

		public void inserirFinal(Inteiro novo){
			Celula novaCelula;

			novaCelula = new Celula(novo);

			ultimo.Proximo = novaCelula;
			novaCelula.Anterior = ultimo;

			ultimo = novaCelula;

			tamanho++;

		}

		public Inteiro removerFinal(){
			Celula removida, penultima;

			if (!listaVazia()){
				removida = ultimo;

				penultima = ultimo.Anterior;
				penultima.Proximo = null;
				removida.Anterior = null;

				ultimo = penultima;

				tamanho--;

				return (removida.Item);
			}
			else
				throw new Exception("Não foi possível remover o último item da lista: a lista está vazia!");
		}

		public void imprimir(){
			Celula aux;

			if (!listaVazia()){
				Console.WriteLine("Conteúdo da lista:");
				aux = primeiro.Proximo;
				while (aux != null){
					aux.Item.imprimir();
					aux = aux.Proximo;
				}
			}
			else
				throw new Exception("Não foi possível imprimir o conteúdo da lista: a lista está vazia!");
		}
	}
}
